Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/MarkerDetection/QRCode/QRCodeMarkerDetector.cs

```diff
@@ -0,0 +1,257 @@
+                        break;
+                    }
+                }
+            }
+
+            return found;
+        }
+
+#if WINDOWS_UWP && QRCODESTRACKER_BINARY_AVAILABLE
+        protected void Start()
+        {
+            if (_qrCodesManager == null)
+            {
+                _qrCodesManager = QRCodesManager.FindOrCreateQRCodesManager(gameObject);
+                _qrCodesManager.QRCodeAdded += QRCodeAdded;
+                _qrCodesManager.QRCodeRemoved += QRCodeRemoved;
+                _qrCodesManager.QRCodeUpdated += QRCodeUpdated;
+            }
+
+            var result = _qrCodesManager.StartQRTracking();
+            Debug.Log("Started qr tracker: " + result.ToString());
+        }
+
+        protected void Update()
+        {
+            if (_tracking &&
+                _processMarkers)
+            {
+                ProcessMarkerUpdates();
+            }
+        }
+
+        private void QRCodeAdded(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            int markerId;
+            lock (_cachedMarkers)
+            {
+                if (!_cachedMarkers.ContainsKey(e.Data.Id) &&
+                    TryGetMarkerId(e.Data.Code, out markerId))
+                {
+                    _cachedMarkers.Add(e.Data.Id, new QRCodeInfo(e.Data.Id, markerId, e.Data.PhysicalSizeMeters));
+                    _processMarkers = true;
+                }
+            }
+        }
+
+        private void QRCodeUpdated(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            int markerId;
+            lock (_cachedMarkers)
+            {
+                if (!_cachedMarkers.ContainsKey(e.Data.Id) &&
+                TryGetMarkerId(e.Data.Code, out markerId))
+                {
+                    _cachedMarkers.Add(e.Data.Id, new QRCodeInfo(e.Data.Id, markerId, e.Data.PhysicalSizeMeters));
+                    _processMarkers = true;
+                }
+            }
+        }
+
+        private void QRCodeRemoved(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            lock(_cachedMarkers)
+            {
+                if (_cachedMarkers.ContainsKey(e.Data.Id))
+                {
+                    _cachedMarkers.Remove(e.Data.Id);
+                    _processMarkers = true;
+                    if (_markerCoordinateSystems.ContainsKey(e.Data.Id))
+                    {
+                        _markerCoordinateSystems.Remove(e.Data.Id);
+                    }
+                }
+            }
+        }
+
+        private void ProcessMarkerUpdates()
+        {
+            lock (_cachedMarkers)
+            {
+                foreach (var markerPair in _cachedMarkers)
+                {
+                    if (!_markerCoordinateSystems.ContainsKey(markerPair.Key))
+                    {
+                        var coordinateSystem = SpatialGraphInteropPreview.CreateCoordinateSystemForNode(markerPair.Key);
```

Comment:
Is there an expectation of acceptable failure of creation? I would imagine it should throw an exception if it fails, and otherwise not return null

Reply:
it currently returns null. This fails fairly frequently and doesn't throw in its current form. 

---
In reply to: [275477203](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/3937#discussion_r275477203) [](ancestors = 275477203)

Reply:
We should address this in the future. Changes will still be needed in this detector class when migrating to the final packaging story of the qrcodesmanager logic.